From a66988b9684447113bd1bfe42a9b5496a577d7fb Mon Sep 17 00:00:00 2001
From: Chris Mi <chrism@mellanox.com>
Date: Mon, 13 Aug 2018 22:20:25 -0400
Subject: [PATCH 2/2] centos 7.5

---
 net.c | 29 +++++++++++++++++++----------
 1 file changed, 19 insertions(+), 10 deletions(-)

diff --git a/net.c b/net.c
index ced5be2..0667e0a 100644
--- a/net.c
+++ b/net.c
@@ -1944,13 +1944,17 @@ cmd_hash(void)
 
 void show_mlx(ulong net_addr)
 {
-        int ofed = 0;
+        int centos7_2 = 0;
+        int centos7_5 = 0;
 	ulong mlx5e_priv = net_addr + SIZE(net_device);
 
 	struct new_utsname *uts;
 	uts = &kt->utsname;
-	if (strncmp(uts->release, "3.10.0-327.el7.x86_64", 6) == 0) {
-		ofed = 1;
+	if (strncmp(uts->release, "3.10.0-327.el7.x86_64", 11) == 0) {
+		centos7_2 = 1;
+		fprintf(fp, "%s\n", uts->release);
+	} else if (strncmp(uts->release, "3.10.0-862.el7.x86_64", 11) == 0) {
+		centos7_5 = 1;
 		fprintf(fp, "%s\n", uts->release);
 	}
 
@@ -1993,7 +1997,7 @@ void show_mlx(ulong net_addr)
 		fprintf(fp, "mlx5_eswitch_rep  %lx\n", rep + STRUCT_SIZE("mlx5_eswitch_rep"));
 		fprintf(fp, "mlx5_eswitch_rep  %lx\n", rep + STRUCT_SIZE("mlx5_eswitch_rep") * 2);
 
-		if (ofed == 0) {
+		if (centos7_2 == 0) {
 			ulong mlx5e_rep_priv = read_pointer2(rep, "mlx5_eswitch_rep_if", "priv");
 			fprintf(fp, "mlx5e_rep_priv  %lx\n", mlx5e_rep_priv);
 			ulong ht = mlx5e_rep_priv + MEMBER_OFFSET("mlx5e_rep_priv", "tc_ht");
@@ -2086,7 +2090,7 @@ void show_mlx(ulong net_addr)
 	fprintf(fp, "netdev_queue  %lx\n", ingress_queue);
 
 	// for centos 7.2
-	if (ofed == 1 && ingress_queue) {
+	if (centos7_2 == 1 && ingress_queue) {
 		fprintf(fp, "for centos 7.2\n");
 		ulong qdisc_sleep = read_pointer2(ingress_queue, "netdev_queue", "qdisc_sleeping");
 		fprintf(fp, "Qdisc %lx\n", qdisc_sleep);
@@ -2113,6 +2117,8 @@ void show_mlx(ulong net_addr)
 	ulong ingress_sched_data = qdisc + STRUCT_SIZE("Qdisc");
 	fprintf(fp, "ingress_sched_data  %lx\n", ingress_sched_data);
 	ulong tcf_block = read_pointer2(ingress_sched_data, "ingress_sched_data", "block");
+	// workaround for centos7.5
+	tcf_block = read_pointer1(ingress_sched_data);
 	if (!tcf_block)
 		return;
 	fprintf(fp, "tcf_block  %lx\n", tcf_block);
@@ -2130,12 +2136,15 @@ void show_mlx(ulong net_addr)
 	fprintf(fp, "list -H %lx -o tcf_chain.list -s tcf_chain\n", chain_list);
 	fprintf(fp, "list tcf_proto.next filter_chain_addr -s tcf_proto\n\n");
 
-	ulong miniq = read_pointer2(net_addr, "net_device", "miniq_ingress");
-	if (!miniq)
-		return;
-	fprintf(fp, "mini_Qdisc  %lx\n", miniq);
 
-	tcf_block = read_pointer2(miniq, "mini_Qdisc", "filter_list");
+	if (!centos7_5) {
+		ulong miniq = read_pointer2(net_addr, "net_device", "miniq_ingress");
+		if (!miniq)
+			return;
+		fprintf(fp, "mini_Qdisc  %lx\n", miniq);
+
+		tcf_block = read_pointer2(miniq, "mini_Qdisc", "filter_list");
+	}
 
 	fprintf(fp, "list tcf_proto.next %lx -s tcf_proto\n", tcf_block);
 
-- 
1.8.3.1

