From af7c49291019cc24feeef15642935c0ddf801fd5 Mon Sep 17 00:00:00 2001
From: Chris Mi <chrism@mellanox.com>
Date: Wed, 11 Dec 2019 10:54:05 +0800
Subject: [PATCH 2/9] change to work_queue

---
 net/openvswitch/vport-netdev.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/net/openvswitch/vport-netdev.c b/net/openvswitch/vport-netdev.c
index d25de4d4abcb..d3b3376c834a 100644
--- a/net/openvswitch/vport-netdev.c
+++ b/net/openvswitch/vport-netdev.c
@@ -39,7 +39,7 @@
 static struct vport_ops ovs_netdev_vport_ops;
 
 static void ovs_netdev_down_event(struct work_struct *dummy);
-static DECLARE_DELAYED_WORK(ovs_netdev_down_work, ovs_netdev_down_event);
+static DECLARE_WORK(ovs_netdev_down_work, ovs_netdev_down_event);
 
 static void ovs_netdev_down_event(struct work_struct *dummy)
 {
@@ -88,7 +88,7 @@ static int netdev_port_receive(struct sk_buff *skb)
 			skb_pull(skb, ETH_HLEN);
 			skb->ip_summed = CHECKSUM_NONE;
 		}
-		mod_delayed_work(system_wq, &ovs_netdev_down_work, 0);
+		queue_work(system_wq, &ovs_netdev_down_work);
 	}
 
 	return ret;
-- 
2.14.4

