From 2e2dd79564692207946f81f403434b66f28da9bc Mon Sep 17 00:00:00 2001
From: Chris Mi <chrism@mellanox.com>
Date: Wed, 11 Dec 2019 13:29:07 +0800
Subject: [PATCH 3/9] last_jiffies

---
 net/openvswitch/vport-netdev.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/net/openvswitch/vport-netdev.c b/net/openvswitch/vport-netdev.c
index d3b3376c834a..1f6ff23af923 100644
--- a/net/openvswitch/vport-netdev.c
+++ b/net/openvswitch/vport-netdev.c
@@ -40,11 +40,17 @@ static struct vport_ops ovs_netdev_vport_ops;
 
 static void ovs_netdev_down_event(struct work_struct *dummy);
 static DECLARE_WORK(ovs_netdev_down_work, ovs_netdev_down_event);
+static unsigned long last_jiffies;
 
 static void ovs_netdev_down_event(struct work_struct *dummy)
 {
 	struct net_device *dev;
 
+	if (jiffies && time_before(jiffies, last_jiffies + HZ * 5)) {
+		last_jiffies = jiffies;
+		return;
+	}
+
 	rtnl_lock();
 	rcu_read_lock();
 	for_each_netdev_rcu(&init_net, dev) {
-- 
2.14.4

