From c897aa34bb25b46617160c83b9679a795d5fc4a9 Mon Sep 17 00:00:00 2001
From: Chris Mi <chrism@mellanox.com>
Date: Tue, 29 May 2018 04:15:13 -0400
Subject: [PATCH 4/5] check for ofed

---
 net.c | 22 ++++++++++++++++------
 1 file changed, 16 insertions(+), 6 deletions(-)

diff --git a/net.c b/net.c
index efc8af3..133c782 100644
--- a/net.c
+++ b/net.c
@@ -1936,9 +1936,17 @@ ulong table_test;
 
 void show_mlx(ulong net_addr)
 {
-        int old = 0;
+        int ofed = 0;
 	ulong mlx5e_priv = net_addr + SIZE(net_device);
 
+	struct new_utsname *uts;
+	uts = &kt->utsname;
+	if (strncmp(uts->release, "3.10.0-327.el7.x86_64", 6) == 0) {
+		ofed = 1;
+		fprintf(fp, "%s\n", uts->release);
+	}
+
+
 	fprintf(fp, "mlx5e_priv  %lx\n", mlx5e_priv);
 
 	/*
@@ -1971,10 +1979,12 @@ void show_mlx(ulong net_addr)
 		fprintf(fp, "mlx5_eswitch_rep  %lx\n", rep + STRUCT_SIZE("mlx5_eswitch_rep"));
 		fprintf(fp, "mlx5_eswitch_rep  %lx\n", rep + STRUCT_SIZE("mlx5_eswitch_rep") * 2);
 
-		ulong mlx5e_rep_priv = read_pointer2(rep, "mlx5_eswitch_rep_if", "priv");
-		fprintf(fp, "mlx5e_rep_priv  %lx\n", mlx5e_rep_priv);
-		ulong ht = read_pointer2(mlx5e_rep_priv, "mlx5e_rep_priv", "tc_ht");
-		fprintf(fp, "hash %lx -s mlx5e_tc_flow -m node\n", ht);
+		if (ofed == 0) {
+			ulong mlx5e_rep_priv = read_pointer2(rep, "mlx5_eswitch_rep_if", "priv");
+			fprintf(fp, "mlx5e_rep_priv  %lx\n", mlx5e_rep_priv);
+			ulong ht = read_pointer2(mlx5e_rep_priv, "mlx5e_rep_priv", "tc_ht");
+			fprintf(fp, "hash %lx -s mlx5e_tc_flow -m node\n", ht);
+		}
 	}
 
 	ulong mdev = read_pointer2(mlx5e_priv, "mlx5e_priv", "mdev");
@@ -2059,7 +2069,7 @@ void show_mlx(ulong net_addr)
 	fprintf(fp, "netdev_queue  %lx\n", ingress_queue);
 
 	// for centos 7.2
-	if (old == 1) {
+	if (ofed == 1 && ingress_queue) {
 		fprintf(fp, "for centos 7.2\n");
 		ulong qdisc_sleep = read_pointer2(ingress_queue, "netdev_queue", "qdisc_sleeping");
 		fprintf(fp, "Qdisc %lx\n", qdisc_sleep);
-- 
2.14.2

