--- kdumpctl.orig	2018-06-27 18:16:10.316374273 +0900
+++ kdumpctl	2018-06-27 18:17:54.860095679 +0900
@@ -481,6 +481,7 @@
 setup_initrd()
 {
 	DEFAULT_INITRD="${KDUMP_BOOTDIR}/initramfs-`uname -r`.img"
+	DEFAULT_INITRD="${KDUMP_BOOTDIR}/initramfs-4.17.2-100.fc27.x86_64.img"
 	DEFAULT_INITRD_BAK="${KDUMP_BOOTDIR}/.initramfs-`uname -r`.img.default"
 	if [ $DEFAULT_DUMP_MODE == "fadump" ]; then
 		TARGET_INITRD="$DEFAULT_INITRD"
@@ -705,6 +706,7 @@
 		kdump_kver=$KDUMP_KERNELVER
 	fi
 
+	kdump_kver=4.17.2-100.fc27.x86_64
 	kdump_kernel="${KDUMP_BOOTDIR}/${KDUMP_IMG}-${kdump_kver}${KDUMP_IMG_EXT}"
 	setup_initrd
 	if [ $? -ne 0 ]; then
@@ -826,9 +828,11 @@
 		KEXEC_ARGS="$KEXEC_ARGS -s"
 	fi
 
+set -x
 	$KEXEC $KEXEC_ARGS $standard_kexec_args \
 		--command-line="$KDUMP_COMMANDLINE" \
 		--initrd=$TARGET_INITRD $kdump_kernel
+set +x
 	if [ $? == 0 ]; then
 		echo "kexec: loaded kdump kernel"
 		return 0
